3D plot GUI plugin

chapulina opened the pull request:
# 🎉 New feature

Closes #231 

Alternative to #332

## Summary
<!--Explain changes made, the expected behavior, and provide any other additional
context (e.g., screenshots, gifs) if appropriate.-->

Plots the path of entities in the 3D scene, parameters are customizable from SDF and / or the GUI.

~~WIP while I work on some final usability tweaks, and possibly a test.~~

Ready for review!

## Test it
<!--Explain how reviewers can test this new feature manually.-->

```
ign gazebo -r plot_3d.sdf
```

This is what it currently looks like:

![plot3d](https://user-images.githubusercontent.com/5751272/125845288-dcf517ef-32d5-4c63-bcd4-cc8abd2627e8.gif)


## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

ahcorde commented on examples/worlds/plot_3d.sdf:
```diff
@@ -0,0 +1,155 @@
+        <play_pause>true</play_pause>
+        <step>true</step>
+        <start_paused>true</start_paused>
+
```
```suggestion
```

chapulina replied:
Just curious, are you following any specific XML style for these suggestions? We haven't been enforcing any blank line policy so far.

ahcorde commented on examples/worlds/plot_3d.sdf:
```diff
@@ -0,0 +1,155 @@
+        <real_time>true</real_time>
+        <real_time_factor>true</real_time_factor>
+        <iterations>true</iterations>
+
```
```suggestion
```

ahcorde commented on include/ignition/gazebo/Util.hh:
```diff
@@ -51,6 +51,14 @@ namespace ignition
       bool _includePrefix = true);
 
+    /// \brief Helper function to get an entity given its scoped name.
+    /// \param[in] _scopedName Entity's scoped name.
```
name doesn't mach

ahcorde commented on src/gui/plugins/plot_3d/Plot3D_TEST.cc:
```diff
@@ -0,0 +1,112 @@
+#include "Plot3D.hh"
+
+int g_argc = 1;
+char **g_argv;
```
it requires a initial value.

chapulina replied:
3c50512

ahcorde reviewed (changes requested):
I made a initial review, I'm going a tried in my machine

chapulina commented on src/gui/plugins/plot_3d/Plot3D.hh:
```diff
@@ -0,0 +1,217 @@
+  /// `<color>`: RGB color of line, defaults to blue.
+  /// `<offset>`: XYZ offset from the entity's origin to plot from, expressed
+  /// in the entity's frame. Defaults to zero.
+  /// `<minimum_distance>`: The minimum distance between points to plot. A new
```
See acd159d

ahcorde commented:
Resize looks weird. X and R are bigger than the other parameters.

![resize](https://user-images.githubusercontent.com/1933907/126665756-1763ec9e-8d61-4c1a-9320-131a8710bb3c.gif)

ahcorde commented on src/gui/plugins/plot_3d/Plot3D_TEST.cc:
```diff
@@ -0,0 +1,115 @@
+  ASSERT_EQ(plugins.size(), 1);
+
+  auto plugin = plugins[0];
+  EXPECT_EQ(plugin->Title(), "Plot3D!");
+  EXPECT_EQ(plugin->TargetEntity(), gazebo::kNullEntity);
+  EXPECT_EQ(plugin->TargetName(), QString("banana"))
+      << plugin->TargetName().toStdString();
+  EXPECT_EQ(plugin->Color(), QVector3D(0.1, 0.2, 0.3));
+  EXPECT_EQ(plugin->Offset(), QVector3D(1, 2, 3));
```
EXPECT_EQ( Expectation, value to test );

azeey reviewed (commented):
Works great! Do you by any chance know why the color of the line changes over time? For example, in your video, the line for the lower arm starts out a light green color and goes to dark green.

chapulina commented on include/ignition/gazebo/Util.hh:
```diff
@@ -51,6 +51,28 @@ namespace ignition
+    /// \param[in] _delim Delimiter between names, defaults to "::".
+    /// \return The engity or kNullEntity if no entity was found with that
+    /// name.
+    Entity IGNITION_GAZEBO_VISIBLE entityFromScopedName(
```
I thought I had mentioned somewhere but now I see it's not here. I targeted Fortress so it's easier to test with another project, but my plan is to retarget it to Citadel after the PR has been reviewed and forward-port it from there :ok_hand:

chapulina commented on include/ignition/gazebo/Util.hh:
```diff
@@ -51,6 +51,28 @@ namespace ignition
+    /// \return The engity or kNullEntity if no entity was found with that
+    /// name.
+    Entity IGNITION_GAZEBO_VISIBLE entityFromScopedName(
+      const std::string &_scopedName, const EntityComponentManager &_ecm,
```
Ah good point, I was living in a world with unique entity names :woman_facepalming: 

Requiring an entity may work well for server plugins that are attached to entities, but GUI plugins don't really have a reference entity, so it would be cumbersome to figure out what entity we're `relative_to`. And I don't like the idea of requiring fully scoped names mainly because it makes it difficult to use the same GUI config with various worlds.

So, I have an alternative suggestion in a1cf2841ea2c57f3ef1e4bad464e76cfe829916b, which is to return all entities that match the scoped name. Then the caller can decide what to do with all of them. Let me know whether that works for your use case. I may have pushed the cumbersomeness to your side :sweat_smile:

azeey replied:
It does indeed become very cumbersome for system writers if they have to sift through a list of entities to determine which entity they should pick. If the issue is figuring out what the world name should be, I think that information is available in `GuiRunner`, so it shouldn't be too difficult to pass it on to `GuiSystem`s. Then, we can have scoped names in GUI configs without needing the world name

chapulina replied:
Fair enough. Added `relativeTo` in 673388d6d058adf8c4dd4d4bc0d89d65c7c82053.

I kept the logic returning multiple entities just in case. Gazebo isn't enforcing unique names, so it isn't an impossibility. I can also imagine this being used to return multiple entities on purpose, for example:

```
model1
  the_link
    visual
model2
  the_link
    visual
```

If I want all visuals from all `the_link`s, I can get them with `entitiesFromScopedName("the_link::visual", ecm)`

chapulina replied:
@azeey , I'll wait on your ok on this function before merging

azeey replied:
Looks good. Thanks!

chapulina commented on src/Util.cc:
```diff
@@ -128,6 +132,46 @@ std::string scopedName(const Entity &_entity,
+    Entity entity;
+    if (kNullEntity == result)
+    {
+      entity = _ecm.EntityByComponents(components::Name(name));
```
Nope, see new implementation

chapulina commented:
Thanks for the reviews, I believe I've addressed all comments :heavy_check_mark:

caguero reviewed (approved):
Looks great!

azeey commented on include/ignition/gazebo/Util.hh:
```diff
@@ -51,6 +51,34 @@ namespace ignition
+    /// the scoped name could be relative to any entity.
+    /// \param[in] _delim Delimiter between names, defaults to "::", it can't
+    /// be empty.
+    /// \return The entity or kNullEntity if no entity was found with that
```
Can you update the `\return` comment so it reflects that an `unordered_set` is returned.

azeey reviewed (approved):
Thanks for iterating. There are a couple of codecheck issues, otherwise LGTM!

chapulina commented:
> Do you by any chance know why the color of the line changes over time? 

Oh, btw, no, I don't. I think this may be an issue on the rendering end though. I'll see if I gather enough to ticket an issue.

chapulina commented:
Retargeted from Fortress to Citadel after iterating:

1. Rebased from `main` to `ign-gazebo3`
2. Updated target branch and label
3. Made necessary changes in 362570b 

If anyone wants to use this feature with Fortress before the PR is forward-ported, checkout this commit: 8025da67c85e07730e26f8cad41a7608a58b0a8e.

I'll merge this into Citadel if CI is happy.
